File: flink-connectors/flink-hcatalog/src/main/scala/org/apache/flink/hcatalog/scala/HCatInputFormat.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.hcatalog.scala

import org.apache.flink.configuration
import org.apache.flink.hcatalog.HCatInputFormatBase
import org.apache.hadoop.conf.Configuration
import org.apache.hive.hcatalog.data.HCatRecord
import org.apache.hive.hcatalog.data.schema.HCatFieldSchema

/**
 * A InputFormat to read from HCatalog tables.
 * The InputFormat supports projection (selection and order of fields) and partition filters.
 *
 * Data can be returned as [[HCatRecord]] or Scala tuples.
 * Scala tuples support only up to 22 fields.
 *
 */
class HCatInputFormat[T](
                        database: String,
                        table: String,
                        config: Configuration
                         ) extends HCatInputFormatBase[T](database, table, config) {

  def this(database: String, table: String) {
    this(database, table, new Configuration)
  }

  var vals: Array[Any] = Array[Any]()

  override def configure(parameters: configuration.Configuration): Unit = {
    super.configure(parameters)
    vals = new Array[Any](fieldNames.length)
  }

  override protected def getMaxFlinkTupleSize: Int = 22

  override protected def buildFlinkTuple(t: T, record: HCatRecord): T = {

    // Extract all fields from HCatRecord
    var i: Int = 0
    while (i < this.fieldNames.length) {

        val o: AnyRef = record.get(this.fieldNames(i), this.outputSchema)

        // partition columns are returned as String
        //   Check and convert to actual type.
        this.outputSchema.get(i).getType match {
          case HCatFieldSchema.Type.INT =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toInt
            }
            else {
              vals(i) = o.asInstanceOf[Int]
            }
          case HCatFieldSchema.Type.TINYINT =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toInt.toByte
            }
            else {
              vals(i) = o.asInstanceOf[Byte]
            }
          case HCatFieldSchema.Type.SMALLINT =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toInt.toShort
            }
            else {
              vals(i) = o.asInstanceOf[Short]
            }
          case HCatFieldSchema.Type.BIGINT =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toLong
            }
            else {
              vals(i) = o.asInstanceOf[Long]
            }
          case HCatFieldSchema.Type.BOOLEAN =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toBoolean
            }
            else {
              vals(i) = o.asInstanceOf[Boolean]
            }
          case HCatFieldSchema.Type.FLOAT =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toFloat
            }
            else {
              vals(i) = o.asInstanceOf[Float]
            }
          case HCatFieldSchema.Type.DOUBLE =>
            if (o.isInstanceOf[String]) {
              vals(i) = o.asInstanceOf[String].toDouble
            }
            else {
              vals(i) = o.asInstanceOf[Double]
            }
          case HCatFieldSchema.Type.STRING =>
            vals(i) = o
          case HCatFieldSchema.Type.BINARY =>
            if (o.isInstanceOf[String]) {
              throw new RuntimeException("Cannot handle partition keys of type BINARY.")
            }
            else {
              vals(i) = o.asInstanceOf[Array[Byte]]
            }
          case HCatFieldSchema.Type.ARRAY =>
            if (o.isInstanceOf[String]) {
              throw new RuntimeException("Cannot handle partition keys of type ARRAY.")
            }
            else {
              vals(i) = o.asInstanceOf[List[Object]]
            }
          case HCatFieldSchema.Type.MAP =>
            if (o.isInstanceOf[String]) {
              throw new RuntimeException("Cannot handle partition keys of type MAP.")
            }
            else {
              vals(i) = o.asInstanceOf[Map[Object, Object]]
            }
          case HCatFieldSchema.Type.STRUCT =>
            if (o.isInstanceOf[String]) {
              throw new RuntimeException("Cannot handle partition keys of type STRUCT.")
            }
            else {
              vals(i) = o.asInstanceOf[List[Object]]
            }
          case _ =>
            throw new RuntimeException("Invalid type " + this.outputSchema.get(i).getType +
              " encountered.")
        }

        i += 1
      }
    createScalaTuple(vals)
  }

  private def createScalaTuple(vals: Array[Any]): T = {

    this.fieldNames.length match {
      case 1 =>
        new Tuple1(vals(0)).asInstanceOf[T]
      case 2 =>
        new Tuple2(vals(0), vals(1)).asInstanceOf[T]
      case 3 =>
        new Tuple3(vals(0), vals(1), vals(2)).asInstanceOf[T]
      case 4 =>
        new Tuple4(vals(0), vals(1), vals(2), vals(3)).asInstanceOf[T]
      case 5 =>
        new Tuple5(vals(0), vals(1), vals(2), vals(3), vals(4)).asInstanceOf[T]
      case 6 =>
        new Tuple6(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5)).asInstanceOf[T]
      case 7 =>
        new Tuple7(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6)).asInstanceOf[T]
      case 8 =>
        new Tuple8(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7))
          .asInstanceOf[T]
      case 9 =>
        new Tuple9(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8)).asInstanceOf[T]
      case 10 =>
        new Tuple10(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9)).asInstanceOf[T]
      case 11 =>
        new Tuple11(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10)).asInstanceOf[T]
      case 12 =>
        new Tuple12(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11)).asInstanceOf[T]
      case 13 =>
        new Tuple13(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12)).asInstanceOf[T]
      case 14 =>
        new Tuple14(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13)).asInstanceOf[T]
      case 15 =>
        new Tuple15(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14)).asInstanceOf[T]
      case 16 =>
        new Tuple16(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15))
          .asInstanceOf[T]
      case 17 =>
        new Tuple17(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16)).asInstanceOf[T]
      case 18 =>
        new Tuple18(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16), vals(17)).asInstanceOf[T]
      case 19 =>
        new Tuple19(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16), vals(17), vals(18)).asInstanceOf[T]
      case 20 =>
        new Tuple20(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16), vals(17), vals(18), vals(19)).asInstanceOf[T]
      case 21 =>
        new Tuple21(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16), vals(17), vals(18), vals(19), vals(20)).asInstanceOf[T]
      case 22 =>
        new Tuple22(vals(0), vals(1), vals(2), vals(3), vals(4), vals(5), vals(6), vals(7),
          vals(8), vals(9), vals(10), vals(11), vals(12), vals(13), vals(14), vals(15),
          vals(16), vals(17), vals(18), vals(19), vals(20), vals(21)).asInstanceOf[T]
      case _ =>
        throw new RuntimeException("Only up to 22 fields supported for Scala Tuples.")

  }

  }
}

File: flink-connectors/flink-hcatalog/src/main/java/org/apache/flink/hcatalog/java/HCatInputFormat.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.hcatalog.java;

import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.hcatalog.HCatInputFormatBase;

import org.apache.hadoop.conf.Configuration;
import org.apache.hive.hcatalog.common.HCatException;
import org.apache.hive.hcatalog.data.HCatRecord;

/**
 * A InputFormat to read from HCatalog tables. The InputFormat supports projection (selection and
 * order of fields) and partition filters.
 *
 * <p>Data can be returned as {@link HCatRecord} or Flink {@link
 * org.apache.flink.api.java.tuple.Tuple}. Flink tuples support only up to 25 fields.
 *
 * @param <T>
 */
public class HCatInputFormat<T> extends HCatInputFormatBase<T> {
    private static final long serialVersionUID = 1L;

    public HCatInputFormat() {}

    public HCatInputFormat(String database, String table) throws Exception {
        super(database, table);
    }

    public HCatInputFormat(String database, String table, Configuration config) throws Exception {
        super(database, table, config);
    }

    @Override
    protected int getMaxFlinkTupleSize() {
        return 25;
    }

    @Override
    protected T buildFlinkTuple(T t, HCatRecord record) throws HCatException {

        Tuple tuple = (Tuple) t;

        // Extract all fields from HCatRecord
        for (int i = 0; i < this.fieldNames.length; i++) {

            // get field value
            Object o = record.get(this.fieldNames[i], this.outputSchema);

            // Set field value in Flink tuple.
            // Partition columns are returned as String and
            //   need to be converted to original type.
            switch (this.outputSchema.get(i).getType()) {
                case INT:
                    if (o instanceof String) {
                        tuple.setField(Integer.parseInt((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case TINYINT:
                    if (o instanceof String) {
                        tuple.setField(Byte.parseByte((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case SMALLINT:
                    if (o instanceof String) {
                        tuple.setField(Short.parseShort((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case BIGINT:
                    if (o instanceof String) {
                        tuple.setField(Long.parseLong((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case BOOLEAN:
                    if (o instanceof String) {
                        tuple.setField(Boolean.parseBoolean((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case FLOAT:
                    if (o instanceof String) {
                        tuple.setField(Float.parseFloat((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case DOUBLE:
                    if (o instanceof String) {
                        tuple.setField(Double.parseDouble((String) o), i);
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case STRING:
                    tuple.setField(o, i);
                    break;
                case BINARY:
                    if (o instanceof String) {
                        throw new RuntimeException("Cannot handle partition keys of type BINARY.");
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case ARRAY:
                    if (o instanceof String) {
                        throw new RuntimeException("Cannot handle partition keys of type ARRAY.");
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case MAP:
                    if (o instanceof String) {
                        throw new RuntimeException("Cannot handle partition keys of type MAP.");
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                case STRUCT:
                    if (o instanceof String) {
                        throw new RuntimeException("Cannot handle partition keys of type STRUCT.");
                    } else {
                        tuple.setField(o, i);
                    }
                    break;
                default:
                    throw new RuntimeException("Invalid Type");
            }
        }

        return (T) tuple;
    }
}
